cluster-autoscaler/cloudprovider/azure/azure_vms_pool.go

```diff
@@ -102,28 +120,265 @@ func (agentPool *VMsPool) MaxSize() int {
+	}
+
+	return agentPool.scaleDownNodes(providerIDs)
+}
+
+// scaleDownNodes delete or deallocate the nodes from the agent pool based on the scale down policy.
+func (agentPool *VMsPool) scaleDownNodes(providerIDs []string) error {
+	agentPool.sizeMutex.Lock()
+	defer agentPool.sizeMutex.Unlock()
+
+	if len(providerIDs) == 0 {
+		return nil
+	}
+
+	klog.V(3).Infof("Deleting nodes from agent pool %s: %v", agentPool.Name, providerIDs)
+
+	deleteCtx, deleteCancel := getContextWithCancel()
+	defer deleteCancel()
+
+	machineNames := make([]*string, len(providerIDs))
+	for i, providerID := range providerIDs {
+		// extract the machine name from the providerID by splitting the providerID by '/' and get the last element
+		// The providerID look like this:
+		// "azure:///subscriptions/0000000-0000-0000-0000-00000000000/resourceGroups/mc_wxrg_play-vms_eastus/providers/Microsoft.Compute/virtualMachines/aks-nodes-32301838-vms0"
+		machineName, err := resourceName(providerID)
+		if err != nil {
+			return err
+		}
+		machineNames[i] = &machineName
+	}
+
+	requestBody := armcontainerservice.AgentPoolDeleteMachinesParameter{
+		MachineNames: machineNames,
+	}
+
+	poller, err := agentPool.manager.azClient.agentPoolClient.BeginDeleteMachines(
+		deleteCtx,
+		agentPool.clusterResourceGroup,
+		agentPool.clusterName,
+		agentPool.Name,
+		requestBody, nil)
+	if err != nil {
+		klog.Errorf("Failed to update agentpool %s in cluster %s with error: %v",
+			agentPool.azureRef.Name, agentPool.clusterName, err)
+		return err
+	}
+
+	defer agentPool.manager.invalidateCache()
+
+	updateCtx, cancel := getContextWithTimeout(updateRequestContextTimeout)
+	defer cancel()
+	if _, err = poller.PollUntilDone(updateCtx, nil); err == nil {
+		return nil
+	}
+
+	klog.Errorf("agentPoolClient.BeginDeleteMachines for aks cluster %s agentpool %s failed with error %s",
+		agentPool.clusterName, agentPool.Name, err)
+
+	return nil
+}
+
+// Belongs returns true if the given k8s node belongs to this vms nodepool.
+func (agentPool *VMsPool) Belongs(node *apiv1.Node) (bool, error) {
+	klog.V(6).Infof("Check if node belongs to this vms pool:%s, node:%v\n", agentPool, node)
+
+	ref := &azureRef{
+		Name: node.Spec.ProviderID,
+	}
+
+	nodeGroup, err := agentPool.manager.GetNodeGroupForInstance(ref)
+	if err != nil {
+		return false, err
+	}
+	if nodeGroup == nil {
+		return false, fmt.Errorf("%s doesn't belong to a known node group", node.Name)
+	}
+	if !strings.EqualFold(nodeGroup.Id(), agentPool.Id()) {
+		return false, nil
+	}
+	return true, nil
 }
 
 // DecreaseTargetSize decreases the target size of the node group.
 func (agentPool *VMsPool) DecreaseTargetSize(delta int) error {
-	// TODO(wenxuan): Implement this method
-	return cloudprovider.ErrNotImplemented
+	agentPool.manager.invalidateCache()
+	_, err := agentPool.getVMsPoolSize()
+	if err != nil {
+		klog.Warningf("DecreaseTargetSize: failed with error: %v", err)
+	}
+	return nil
```

Comment:
Functionally NOOP implementation deserves a comment

Reply:
Oops I'm supposed to return err here, updated the code. Thanks for pointing it out!

Reply:
Yes, but my point is a little different. 

This implementation of `DecreaseTargetSize` (admittedly, same as current implementation for VMSS) does not actually do anything (except invalidating the cache), and yet (essentially) never fails. This is a latent problem, as it is used by autoscaler core to adjust the size of nodepool that appears to need adjusting, and not returning an error makes it think adjustment was made, which can in turn lead to persistent issues (that have been observed). It is not a problem if `DecreaseTargetSize` never gets called (which is the most common case for well-behaving nodepool). But a correct implementation would likely a) actually adjust the target size (subject to constraints in comments for DecreaseTargetSize) and b) return an error if it is not possible or not successful. Not doing this is a choice (or a bug?) - that's what I meant by NOOP implementation deserves a comment.

We likely need to revisit this for VMSS version as well. It does have an explanation for not doing anything in comment ("VMSS size should be changed automatically after the Node deletion, hence this operation is not required."), which I don't fully understand/agree with.

So please give some thought to whether a better implementation of `DecreaseTargetSize` for VM node pool (respecting constraints outlined in comments) is possible. For expediency, it is probably Ok to leave it as is for now (maybe at least with a comment like "just following VMSS implementation"), to revisit later. If you do, would be good to validate in testing that it actually never gets called.